use statically_known_true instead of guard_size_oblivious in pattern matcher

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147557

We shouldn't add guards here. Use statically_known_true instead. Internal xref: https://fb.workplace.com/groups/1075192433118967/?multi_permalinks=1609560723015466&comment_id=1610040026300869&notif_id=1740082892544333&notif_t=work_feedback_reaction_generic&ref=notif

Differential Revision: [D69950122](https://our.internmc.facebook.com/intern/diff/D69950122/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

bobrenjc93 committed:
use statically_known_true instead of guard_size_oblivious in pattern matcher

We shouldn't add guards here. Use statically_known_true instead

Differential Revision: [D69950122](https://our.internmc.facebook.com/intern/diff/D69950122/)

[ghstack-poisoned]

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -1342,7 +1342,7 @@ def check_fn(match: Match) -> bool:
                     for v in itertools.chain(args[i].shape, args[i].stride()):
                         if isinstance(v, torch.SymInt) and all(
-                            guard_size_oblivious(v != a) for a in sym_args
+                            statically_known_true(v != a) for a in sym_args
```
this feels like something has gone wrong in our apis if guard_size_oblivious does not work here. this would be the 3rd iteration of this same code.

ColinPeppler replied:
Hi @eellison, noob q: what could be wrong in the apis, and why is it reasonable to add inequality guards against each symbol here?

eellison replied:
I guess I don't have a good understanding for when we should be using guard size oblivious vs statically known true. I agree, this change seems right. But more generally - when should we be using one vs the other ? and do we need to audit our existing usages.

eellison replied:
ctrl-F guard_size_oblivious gives 345 usages outside of tests where as statically_known_true only gives 58. how many of these are wrong ? cc @bobrenjc93

bobrenjc93 replied:
@laithsakka and I have been discussing this recently and the general conclusion we have is people (especially in fbcode) add guard_size_oblivious without thinking through the implications.

My current mental model is you use `guard_size_oblivious` if it's ok to specialize by adding guards using the hints. In contrast you use `statically_known_true` if you don't want to add guards and are ok to take the other control flow path even if the hint wouldn't have.

bobrenjc93 replied:
@eellison looking at the history, I only see two modifications

D53498537 - which originally introduced guard_size_oblivious
D52718196 - which previously eagerly guarded on this 

Do you know if there were other attempts to write this as `statically_known_true` that thrashed?

eellison replied:
That's fair . Still, I think it would be useful to do a complete audit of guard_size_oblivious compared to statically_known_true, all of the 384+ usages. as opposed to one by one as an issue appears which is what we're doing now. 

i guess basically the heuristic we're trying to do is if there is a compelling enough perf reason use statically_known_true, or maybe, if we need to add a guard in order to get the correct semantics (and maybe we're incorrect on dim 0, 1 but we dont have a better way right now). would you say that's correct ?

Especially for AOTI, I don't know that we ever have a compelling enough perf reason to make it so that we can't export or add an extra guard.

bobrenjc93 replied:
> I think it would be useful to do a complete audit of guard_size_oblivious compared to statically_known_true, all of the 384+ usages

100% agree. I think in addition to the auditing, we may want to actually change the API to be easier to reason about as well.

>  as opposed to one by one as an issue appears which is what we're doing now.

So just to be clear I only put up this PR in response to a user complaint about recompilation. I agree we should do the audit, but I also think incremental progress, especially to unblock end users, is reasonable. Do you think it's ok to land this PR as is? I just worry the audit will take a while, and users in the meantime are stuck in limbo.

eellison replied:
Yes, I think it's fine to land as is. We can do an audit afterward. just dont want to wait for each call site bc of a user issue.

bobrenjc93 committed:
Update on "use statically_known_true instead of guard_size_oblivious in pattern matcher"


We shouldn't add guards here. Use statically_known_true instead. Internal xref: https://fb.workplace.com/groups/1075192433118967/?multi_permalinks=1609560723015466&comment_id=1610040026300869&notif_id=1740082892544333&notif_t=work_feedback_reaction_generic&ref=notif

Differential Revision: [D69950122](https://our.internmc.facebook.com/intern/diff/D69950122/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge

bobrenjc93 commented:
@Pytorchbot merge

bobrenjc93 commented:
@pytorchbot merge

bobrenjc93 commented:
@pytorchbot rebase -b main

bobrenjc93 committed:
Update

[ghstack-poisoned]
